test/integration/gateways/sagepay_test.exs~

```diff
@@ -0,0 +1,125 @@
+defmodule Gringotts.Integration.Gateways.SagePayTest do
+  # Integration tests for the Sagepay 
+
+  use ExUnit.Case, async: false
+  alias Gringotts.Gateways.SagePay
+  alias Gringotts.{CreditCard, Response}
+
+  @card %CreditCard{
+    number: "4484000000002",
+    month: 3,
+    year: 20,
+    first_name: "SAM",
+    last_name: "JONES",
+    verification_code: "123",
+    brand: "VISA"
+  }
+
+  @amount Money.new(100, :GBP)
+
+  @bad_card %CreditCard{
+    number: "4929000005559",
+    month: 3,
+    year: 20,
+    first_name: "SAM",
+    last_name: "JONES",
+    verification_code: "123",
+    brand: "VISA"
+  }
+
+  @address %{
+    address1: "407 St. John Street",
+    city: "London",
+    postalCode: "EC1V 4AB",
+    country: "GB"
+  }
+
+  @opts [
+    config: %{
+      auth_id:
+        "aEpZeHN3N0hMYmo0MGNCOHVkRVM4Q0RSRkxodUo4RzU0TzZyRHBVWHZFNmhZRHJyaWE6bzJpSFNyRnliWU1acG1XT1FNdWhzWFA1MlY0ZkJ0cHVTRHNocktEU1dzQlkxT2lONmh3ZDlLYjEyejRqNVVzNXU=",
+      vendor: "sandbox"
+    },
+    transactionType: "Deferred",
+    description: "Demo Payment",
+    customerFirstName: "Sam",
+    customerLastName: "Jones",
+    billingAddress: @address
+  ]
+
+  @bad_opts [
+    config: %{
+      auth_id:
+        "aEpZeHN3N0hMYmo0MGNCOHVkRVM4Q0RSRkxodUo4RzU0TzZyRHBVWHZFNmhZRHJyaWE6bzJpSFNyRnliWU1acG1XT1FNdWhzWFA1MlY0ZkJ0cHVTRHNocktEU1dzQlkxT2lONmh3ZDlLYjEyejRqNVVzNXU=",
+      vendor: "sandbox"
+    },
+    transactionType: "Deferred",
+    vendorTxCode: "demotransaction-20",
+    description: "Demo Payment",
+    customerFirstName: "Sam",
+    customerLastName: "Jones",
+    billingAddress: @address
+  ]
+
+  @payment_id "699BB561-07B4-FB3E-B695-B64BEAF28D50"
+
+  @moduletag :integration
+
+  setup_all do
+    Application.put_env(
+      :gringotts,
+      Gringotts.Gateways.Sagepay,
+      # some_key: "some_secret_key"
+      []
+    )
+  end
+
+  # Group the test cases by public api
+  describe "purchase" do
+  end
+
+  describe "authorize" do
+    test "successful response with valid params" do
+      random_code = Enum.random(1_000_000..10_000_000000000000) |> Integer.to_string()
```

Comment:
All tests in this block require a `random_code`. You can write a `setup` function for that:
```elixir
defp vendor_tx_code do
  random_code = Enum.random(1_000_000..10_000_000000000000) |> Integer.to_string()
  [random_vendor_code: [vendorTxCode: "demotransaction-" <> random_code]]
end

describe "authorize" do
  setup :vendor_tx_code
  
  test "whatever", %{random_vendor_code: code} do
     opts = @opts ++ code
     ...
  end

  test "whatever", %{random_vendor_code: code} do
     opts = @opts ++ code
     ...
  end
end
``` 

For more details see the ExUnit docs on setup blocks.